File: c++/0111.cpp
// Link: https://leetcode.com/problems/solving-questions-with-brainpower/description/
// Runtime: 286 ms
// Memory: 144.7 MB

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution 
{
public:
    int minDepth(TreeNode* root) 
    {
        if (root == nullptr) return 0;
        std::queue<std::pair<TreeNode*, int>> nodesAndDepths;
        nodesAndDepths.push(std::make_pair(root, 1));

        int minimumDepth = 0;
        bool isFound = false;
        while ((!nodesAndDepths.empty()) && (!isFound))
        {
            TreeNode* node = nodesAndDepths.front().first;
            int depth = nodesAndDepths.front().second;
            nodesAndDepths.pop();

            if (node->left != nullptr) nodesAndDepths.push(std::make_pair(node->left, depth + 1));
            if (node->right != nullptr) nodesAndDepths.push(std::make_pair(node->right, depth + 1));

            if ((node->left == nullptr) && (node->right == nullptr)) 
            {
                minimumDepth = depth;
                isFound = true;
            }
        }

        return minimumDepth;
    }
};

File: python/0111.py
# Link: https://leetcode.com/problems/minimum-depth-of-binary-tree/description/
# Runtime: 502 ms
# Memory: 51.2 MB

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
import queue


class Solution:
    def minDepth(self, root: Optional[TreeNode]) -> int:
        if root == None:
            return 0
        
        nodes_and_depth = queue.Queue()
        nodes_and_depth.put((root, 1))
        is_found = False
        minimum_depth = 0
        while (not nodes_and_depth.empty()) and (not is_found):
            node, depth = nodes_and_depth.get()

            if node.left != None:
                nodes_and_depth.put((node.left, depth + 1))
            if node.right != None:
                nodes_and_depth.put((node.right, depth + 1))
            
            if node.left == None and node.right == None:
                minimum_depth = depth
                is_found = True
        
        return minimum_depth
